### ISSUE TITLE
Add "WebAPI" type

### ISSUE THREAD

**Issue by IvanGoncharov** [schemaorg/schemaorg#1423]
At the moment Schema.org don't have appropriate type to describe Web APIs.
By WebAPI I mean any API which uses HTTP/HTTPS as transport mechanism including REST API, SOAP, GraphQL, etc.
At the moment there is no scalable mechanism to discover publicly available APIs on the internet.
Closest solution so far is API catalogs, and for the last two year, I have been maintaining [my own](https://APIs.guru).
But it's hard to maintain and impossible to scale.

On the other hand, almost all public APIs have dedicated HTML pages where they can use Schema.org types.
Ideally it should have fields like:
  - Title, Description, Logo
  - License, ToS
  - Support contacts
  - Version, e.g. `1.0.0`
  - list of base URLs, e.g. http://api.example.com/v1, https://api.example.com/v1/
  - Type of service: RESTful, SOAP, GraphQL
  - Link to [API Description](https://en.wikipedia.org/wiki/Overview_of_RESTful_API_Description_Languages)

I want to work on a PR adding `WebAPI` type to `schema.org`.
Can you please advice which base type should I derive it from?
And should I submit it as PR or you have some formal procedure?

**Comment by jaygray0919** [comment]
This would complement work done here:
  http://webdatacommons.org/structureddata/

**Comment by IvanGoncharov** [comment]
@jaygray0919 It will benefit entire developers community since a huge amount of data can be accessed only through Web APIs. And for many companies describing their existing Web APIs will be the first step toward Linked Data.

**Comment by IvanGoncharov** [comment]
@danbri @RichardWallis I'm ready to start working on PR, can you please answer my questions?
> Can you please advice which base type should I derive it from?
> And should I submit it as PR or you have some formal procedure?

**Comment by RichardWallis** [comment]
@IvanGoncharov apologies for being a bit slow in responding.

I can see the basic need for something like this - I would like to see some indication of who would apply it to their sites (other than yourself of course).

Most of the properties you describe could be generally available via any Type in Schema (Title (name), description, URLs, etc.) and via the organization Offering the API (Service?) (Logo, Contacts etc.)

As to a base type to enhance or subtype, my inclination is to look at [EntryPoint](http://schema.org/EntryPoint) which is generally describing something similar.  Not sure if we just enhance that type with a few more specific properties, or create a more specific subtype for API (do we need to constrain it to WebAPI ?)

In the current structure EntryPoint is accessed via the target property of [Action](http://schema.org/Action) and its subtypes.    Maybe a new _APIAction_ subtype would work here.  This would enable that a description of the Organization/Service that is providing the API to be linked to the actual API description via potentialAction.

**Comment by thadguidry** [comment]
@RichardWallis I've pinged David and Wendell from ProgrammableWeb.com to voice their opinions here as well.  Hopefully they respond shortly.

**Comment by dberlind** [comment]
Hey all, I've received @thadguidry's email and replied. As you can imagine, this is an issue that is core to our mission. It's practically our purpose in life. So, we do spend a disproportionate amount of our time thinking about API discovery and search and developing an understanding of what exactly developers want when they start to look for APIs. So, we certainly appreciate the outreach to better understand what role we can play.

**Comment by IvanGoncharov** [comment]
> I can see the basic need for something like this - I would like to see some indication of who would apply it to their sites (other than yourself of course).

@RichardWallis I think it's most important question and should be answered before technical ones.
And it should be answered by API owners and DX guys so I publish a post on our blog asking for feedback: https://blog.apis.guru/api-discovery-can-we-do-better-2336706d5407

**Comment by tlrobinson** [comment]
This is a great idea.

I'm not sure "web" is the best name for the type of APIs you're talking about, basically 3rd party APIs. HTTP is certainly the most common transport these days but that might not always be the case. Also, "web" generally implies hypertext (of course that's the "H" in HTTP but almost no APIs actually use hypertext)

**Comment by pjz** [comment]
...so 'net API' instead of 'web API'?

**Comment by shashanksingh** [comment]
Hey Guys,
We at faasos (on-demand-delivery)-would love to have these openly discoverable api schema for our public api's.

URL : www.faasos.com

**Comment by mesteche** [comment]
> I'm not sure "web" is the best name for the type of APIs you're talking about, basically 3rd party APIs. HTTP is certainly the most common transport these days but that might not always be the case. Also, "web" generally implies hypertext (of course that's the "H" in HTTP but almost no APIs actually use hypertext)
With WebSockets and WebRTC, this is becoming less and less true.

**Comment by maccman** [comment]
@clearbit would love to see this.

**Comment by sotaan** [comment]
Hey!
I really like this idea since it will make API discovery smoother

**Comment by L4in** [comment]
As someone who already lost countless hours trying to find a good MIDI API (still haven't, so if you know something good in Python, please tell me), I totally back up this idea.
A lot of people not in tech complained about how programming looks complex by lack of documentation or visibility - I tend to disagree because this is less and less true - and this is definitely a change needed to welcome even more people into writing their own stuff.

**Comment by asbjornu** [comment]
This sounds a lot like [Hydra](http://www.hydra-cg.com/)'s [`ApiDocumentation`](http://www.hydra-cg.com/spec/latest/core/#discovering-a-hydra-powered-web-api). Please have a look; perhaps there's something that could be learned one way or another.

@tlrobinson:
> Nitpick: I'm not sure "web" is the best name for the type of APIs you're talking about, basically 3rd party APIs. HTTP is certainly the most common transport these days but that might not always be the case. Also, "web" generally implies hypertext (of course that's the "H" in HTTP but almost no APIs actually use hypertext)

But won't you agree that having a document describing your API and perhaps even being so bold as to **linking** to your initial API endpoint, would make at least the Web API description a hypertext format?

**Comment by ncthbrt** [comment]
We (WhereIsMyTransport) would be very interested.

**Comment by cmerrick** [comment]
We (stitchdata.com) extract data from dozens of APIs.  We think this is a great idea. +1

**Comment by adamretter** [comment]
+1

**Comment by bert2002** [comment]
Would love to see this happen.

**Comment by mateomurphy** [comment]
EntryPoint would seem to be a good type for individual API calls (which should definitely have their own type). But overall API should probably be a Service. Net and/or Web API could be subtype of API

**Comment by dberlind** [comment]
On ProgrammableWeb, we are making reference more and more to "networkable"
APIs vs. just Web (HTTP) APIs. The latest data model behind our directories
reflects this and one of the strongest use cases has to do with the
Internet of Things which for some "things" doesn't involve HTTP at all (ie:
MQTT over TCP/IP).

**Comment by ivan-gammel** [comment]
I like this proposal and my company will eventually support it by publishing some of our REST APIs this way. Some things I'd personally be glad to see:
1. Indication of REST API maturity level (resources/verbs/HATEOAS)
2. For Level 3 REST APIs what are the supported formats (HAL, Siren etc)

**Comment by dberlind** [comment]
Ivan, how does data get curated or policed? Is it essentially crowdsourced
by populating a schema as is the case with other schemas such as recipes,
reviews, etc. The reason I ask is that, in our experience of doing this
sort of thing for over a decade, we know for sure that many API providers
(not all, but enough to poison the well) will provide whatever data makes
them look best as opposed to what data is most truthful. For example, API
providers will say they support hypermedia when they don't. That sort of
thing. Sometimes, it's because a non-technical person (eg: a marketing or
SEO person) is responsible for maintaining that data.

**Comment by chrisgandi** [comment]
This is really needed

**Comment by schwartzMS** [comment]
I would be interested in implementing this on our site.

**Comment by berezovskyi** [comment]
@dberlind I think the only reliable way to verify that the API conforms to its declaration is to have the API provide this information itself (e.g. https://www.w3.org/TR/ldp-primer/#filelookup for LDP-enabled apps, WSDL for SOAP etc.) and verify it programmatically. But this seems not to be the purpose behind this PR: it would be good to annotate the existing (i.e. legacy) APIs, in a noninvasive manner (i.e. via such tags in the API documentation).

**Comment by dberlind** [comment]
Thanks @andrew. What about subjective information that less easy to verify
and potentially more about gaming search? For example, referring to @Ivan's
Medium post, if I'm an unscrupulous SEO (those don't exist, right?) or a
marketing person who is judged on eyeballs, and I know that the search term
"email API" is getting a lot of hits on Google, I might gladly put "email"
into the category field and description field (assuming those fields are
part of the schema) even if my API isn't exactly an email API.  This is
exactly the behavior we've seen when crowdsourcing this data set on
ProgrammableWeb and we have wrestled with this problem for years. Also,
would a schema be deemed invalid (iow, not processable) if it was missing
the above suggested declaration? If the answer to that is "no," then
ultimately, you still have the issue of search results being gamed or
corrupted.

**Comment by digitalmaster** [comment]
@IvanGoncharov This type of out of the box, community driven thinking is why i love what i do. Would love to see this implemented. 👍

**Comment by ivan-gammel** [comment]
@dberlind to defend from "dark" SEO, it may be useful to include "dependsOn" attribute to API and web site metadata, that will list all APIs referenced by the described resource. API providers will require the inclusion in dependsOn in their terms of service and search engines may use it to rank search results by popularity. API catalogs may also include the APIs that they list and have validated in their dependsOn attribute as a proof of the metadata correctness.

**Comment by dberlind** [comment]
thanks @ivan-gammel, I apologize, but your answer shot right over my head. For example, why would an API provider require "dependsOn in **their** ToS?" Algorithmically, how would "dependsOn" influence rank?

**Comment by berezovskyi** [comment]
@dberlind you're absolutely right and I will not suggest any improvements (also because you've been doing this for a long while to know better than most of us).

However, I think I am not fully convinced by the need of mechanisms to prevent abuse on the API description level. I see this more as HTML vs. Google problem. Anybody can publish a freeform HTML page (including intentionally wrong schema.org markup). Google decides which site goes high in SERP and which one is banned. Therefore, I see the value of API markup in being able to have the lower common denominator accessible to as many developers as possible, while it's @dberlind's job at ProgrammableWeb to make sure only the relevant APIs end up on the front page (thank you for your hard work!).

**Comment by dberlind** [comment]
thanks @berezovskyi..thanks very much for the compliment (I'll be sure to pass it on to our API search team)!  But, for the record, I'm not necessarily looking to be right. For a variety of reasons (as you can imagine), we at ProgrammableWeb are very interested in the idea of machine driven solutions that yield equal or better quality than what we are achieving manually today.  I also agree with the idea of "lowest common denominator" discussions. What's required. What's not. We based a lot of that thinking on the requests we get. For example, why can't I search this way? Or that way? Or, I just want a total of all APIs that match X description.

**Comment by jancurn** [comment]
This would be great, we'd definitely use it on our website https://www.apifier.com/

**Comment by ivan-gammel** [comment]
@dberlind this isn't well-thought idea, I've just tried to propose a mechanism similar to Google page rank. The "dependsOn" is a tool for a "white" SEO: you publish the API, because you want to share it or to sell it, so you want to make it appear credible and reliable. Thus, you ask users to include it in their "dependsOn" metadata same way, as some library authors want to put their names in copyright notices (I think such copyright notices already exist on some web sites, saying something like "data provided by example.com"). When a search engine discovers such API, it can assign a rank, based on number of references to it. Simple algorithm will just use number of "dependsOn" references as a rank. The more API is being used by others, the higher it appears in search results. More advanced ones will also add weight to references, based on the rank of the referencing API or web site, so that just one authoritative user may push the API to the top, increasing chances for new, but really good APIs.

**Comment by IvanGoncharov** [comment]
@dberlind @ivan-gammel Let's stick with the topic of this issue. "dark" SEO is a problem of search engine and is not related exclusively to APIs. This issue is about figuring out how to describe API similar to other types in Schema.org and not invent a new mechanism. 

**Baseline: We are here to discuss a technical solution on how to add a new WebAPI type.** If you want to discuss  Schema.org or SEO or data collection in general [here](https://news.ycombinator.com/item?id=13004833) is the good place to do so.

**Comment by chaals** [comment]
@IvanGoncharov I think "dark" SEO is a reasonable part of this topic. If we get stuck with a design that means search engines are getting bad data, it will mean we don't use it at all. That would penalise the people who went to the trouble of doing it right, and schema.org as a whole would suffer.

So I don't think we should be discussing the general issues of SEO and so on, but I think we need to consider it to the extent that it relates to designing this property.

**Comment by IvanGoncharov** [comment]
> I can see the basic need for something like this - I would like to see some indication of who would apply it to their sites (other than yourself of course).

@RichardWallis  I didn't expect my blog post to become so popular so it's clear indication that it address some of the pain points for many developers. Moreover @shashanksingh @ncthbrt @cmerrick @schwartzMS @jancurn said they will use this type to describe their APIs.
Can we count this question to be resolved and switch to a technical discussion on how this type should be implemented?

**Comment by asbjornu** [comment]
@chaals: The "dark SEO" issue is going to solve itself if people start using hypertext. That's how Google created page rank in the first place. If people stop creating RPC-like HTTP services and start using the "H" in HTTP to create truly RESTful services, these services will have links to and from each other, creating a web of APIs that is going to have the exact same quality metrics as the HTML based web.

If preventing "dark SEO" is what's needed to get people to implement hypermedia in their APIs, then I'm especially against doing anything about it in the `WebAPI` term of schema.org. JSON-LD and other hypermedia formats have "solved" this already. We don't need any more solutions, we just need to stop building dead-end RPC-style APIs and start building hypermedia-driven, RESTful ones.

**Comment by RichardWallis** [comment]
@IvanGoncharov yes I see you have identified quite a community of potential users!

Let's move on potential proposals for how we enhance the vocabulary to satisfy the need.

I refer to my previous comments https://github.com/schemaorg/schemaorg/issues/1423#issuecomment-261529686 for a suggested approach.

**Comment by dberlind** [comment]
@IvanGoncharov I don't participate in a lot of threads on Github, so please accept my apologies if I sullied the thread. I'll refrain from further non-technical comment.

**Comment by torikulhsn** [comment]
Nice site dude. With google adsense http://muviza.review <a href="http://Muviza.review"></a>  
YouSuvee.com http://www.yousuvee.com”></a>

**Comment by shiondev** [comment]
We would love this at [Datafiniti](https://datafiniti.co)

**Comment by chaals** [comment]
[NB: I deleted a comment that was spam]

@asbjornu no, dark SEO doesn't go away because people use HTTP, and pagerank is vulnerable to it - hence the spam comments even here, that I deleted: Because they are a tool specifically taking advantage of "pure pagerank" SEO - which is why search engines rely on more than just pagerank in their algorithms.

The point isn't that we need to block dark SEO in order to enable Web APIs. It is that if we create a mechanism for labeling Web APIs (or anything else) that makes it easier to do dark SEO, then search engines - the primary consumers and motivators of schema.org data - will not use it unless we have the tools to counter dark SEO techniques. And that would make the work far less useful to everyone.

**Comment by IvanGoncharov** [comment]
@RichardWallis 

> Let's move on potential proposals for how we enhance the vocabulary to satisfy the need.
> I refer to my previous comments #1423 (comment) for a suggested approach.

Awesome! 🎉  🎆

> do we need to constrain it to WebAPI ?

Let's define the scope of this proposal. Non-network APIs can be documented with [APIReference](http://schema.org/APIReference).

Documentation for private, partner-only APIs is typically not crawled by search engines bots.

As for IoT APIs, they are different use case since they are typically served by individual devices with the huge number of API versions  (HW model & SW version). So API discovery should be handled on a per-device basis.

So we have left only public APIs hosted by API owners and in the absolute majority they are served via:
 - the internet as a network
 - HTTP (including WebSockets) as transport
 - URL as single entry point or multiple URLs to represent endpoints

Which is very close to `WWW` definition. But since `WWWAPI` looks very confusing  I propose to stick with `WebAPI` which is almost universally recognized by Web developers. Moreover, Wikipedia seems to agree: https://en.wikipedia.org/wiki/Web_API

**Comment by lanthaler** [comment]
I've created PR #1445 to push this forward by having something concrete to discuss. Quickly summarized I think it is a great idea to add a type like `WebAPI` to schema.org but we should keep it as simple as possible initially and defer most of the complexity to the `documentation` a `WebAPI` points to. [More details and thoughts can be found in the pull requests description](https://github.com/schemaorg/schemaorg/pull/1445).

**Comment by danbri** [comment]
Thanks all, and @lanthaler for the PR. It's available under "pending" for further review/discussion, but seems to be taking shape nicely:

See http://pending.webschemas.org/WebAPI

**Comment by lanthaler** [comment]
Great. Asked on Twitter for feedback https://twitter.com/MarkusLanthaler/status/819185316422709248

**Comment by LukasRos** [comment]
The apis.json format can be used to describe APIs that a site offers: http://apisjson.org/index.html
According to the specification an apis.json file can be autodiscovered either through a well-known location (`domain/apis.json`) or through `<link rel="api" ..`; then apis.json can again link to e.g. Swagger for more detailled documentation for each APIs.
Just wanted to point this out since it solves a similar purpose as to not reinvent the wheel.

**Comment by rektide** [comment]
I can see very few straightforward tactics for a number of different API implementers to get together, agree to an API, and declare that they implement that API. We can see examples of this problem recently with the Beaker (Web) Browser, which wants to allow peers in a p2p network to offer services, and needs a way to allow these peers to declare that they offer implementations of well known services: https://twitter.com/pfrazee/status/816086651671822336

It's great to have sites declaring their API, but as a consumer, someone looking for data, what I want to know is who can give me the API that I'm looking for right now.

On Twitter Markus [has suggested](https://twitter.com/MarkusLanthaler/status/816426875086114818)

> sure, that's just the simplest option. another one would be to, e.g., compare docs for (partial) equivalence

That's not impossible but it certainly seems like an onerous path to ask everyone wanting to implement APIs to take. It also allows the potential for clashes, differently behaving APIs.

I don't have a strong well thought out suggestion for the way forward. Perhaps an "implements" property could allow peers to declare a URL of what it is that they are implementing?

**Comment by lanthaler** [comment]
Thanks @LukasRos. APIs.json and numerous other, similar projects are indeed similar. This proposal is about introducing the concept of a Web API as first-class citizen into Schema.org. Describing the technical details of a Web API is explicitly out of scope for this issue. Instead, `WebAPI` points to a `documentation` which can describe the API in all its details. That could be a Apis.json or Swagger file.

> It's great to have sites declaring their API, but as a consumer, someone looking for data, what I want to know is who can give me the API that I'm looking for right now.

@rektide let's try to get there step by step. This proposal is a first step in that direction.

**Comment by JervenBolleman** [comment]
In the example given I miss the URl of the API itself.  For example could one say?
```
<script type="application/ld+json">
{
    "@id":"http://www.uniprot.org/blast",
    "@context": "http://schema.org/",
    "@type": "WebAPI",
    "name": "UniProt BLAST",
    "description": "BLAST against the UniProt databases",
    "documentation": "https://www.uniprot.org/help/blast",
    "provider": {
      "@type": "Organization",
      "name": "UniProt consortium"
    }
}
</script>```

Or should it be different. That "@id" is a url of a webpage, as much as the target of a form and entrypoint of the API.

**Comment by LukasRos** [comment]
@JervenBolleman: By "URI of the API itself", do you mean the endpoint where the API is hosted? The way I understood it is that ´documentation´ should point to a _machine-readable_ documentation e.g. in Swagger/OpenAPI format and _that_ specification would contain the endpoint.

**Comment by RichardWallis** [comment]
I would assume you would use _url_:
```<script type="application/ld+json">
{
    "@id":"http://www.uniprot.org/blast",
    "@context": "http://schema.org/",
    "@type": "WebAPI",
    "name": "UniProt BLAST",
    "description": "BLAST against the UniProt databases",
    "documentation": "https://www.uniprot.org/help/blast",
    "provider": {
      "@type": "Organization",
      "name": "UniProt consortium"
    },
    "url": "http://www.uniprot.org/blast/api"
}
</script>```

_Note: I do not know if that is the correct url for the endpoint._

**Comment by JervenBolleman** [comment]
The point I am trying to make is the example does not give an identity to the API. In RDF terms it is a BlankNode. Assume there are two pages describing an api hosted and maintained by different people. How can they say that they are talking about the same third party api? 

If this is pointer to an API documentation I think the modelling should be a bit different e.g. taking the google knowlegraph example see that I removed `"documentation"` and moved it to `"@id"` and changed the `"@type`"

```json+ld
{
  "@context": "http://schema.org/",
  "@id" :  "https://developers.google.com/knowledge-graph/",
  "@type": "WebAPIDocumentation",
  "name": "Google Knowledge Graph Search API",
  "description": "The Knowledge Graph Search API lets you find entities in the Google Knowledge Graph. The API uses standard schema.org types and is compliant with the JSON-LD specification.",
  "termsOfService": "https://developers.google.com/knowledge-graph/terms",
  "provider": {
    "@type": "Organization",
    "name": "Google Inc."
  }
}
```
In otherwords if this Type is pointing to documentation that is not parseable by just schema.org i.e. because it is a WSDL or plain JSON swagger then the schema.org should identify the documentation and that it is documentation not saying it is the API.

Then a WebAPI as intangible would be replaced by WebAPIDocumentation/Page as child of WebPage or perhaps better.

It would also be more consistent as SoftwareApplication is a creativework, while WebAPI is an intangible.

**Comment by thadguidry** [comment]
@JervenBolleman 

In Richard's example...the property "documentation": "https://www.uniprot.org/help/blast",

could be further explained with additional Schema markup as your example.  We're just giving a simple URL property example...but it can also expect a Type of http://schema.org/APIReference


```
"@type": "WebAPI"
"documentation": {
   "@context": "http://schema.org/",
   "@id" :  "https://developers.google.com/knowledge-graph/",
   "@type": "APIReference",
   "name": "Google Knowledge Graph Search API",
   "description": "The Knowledge Graph Search API lets you find entities in the Google Knowledge Graph. The API uses standard schema.org types and is compliant with the JSON-LD specification.",
   "termsOfService": "https://developers.google.com/knowledge-graph/terms",
   "provider": {
     "@type": "Organization",
     "name": "Google Inc."
  }
}
```

**Comment by LukasRos** [comment]
Ah, I think I got your point, thanks for clarifying. As per my understanding of schema.org it typically deals with blank nodes, none of the examples on schema.org have an `@id` (the fact that a page contains a JSON-LD, RDFa or Microdata snippet implicitly says "this page is about this object" without giving the object a unique identifier; as such an identifier may not even exist).

So, yes, I think there should be `@id` to give the API a unique identifier, just as in @JervenBolleman's [first example from yesterday](https://github.com/schemaorg/schemaorg/issues/1423#issuecomment-284719368).

That identifying URI may be identical to the API endpoint, the documentation page or even something else (e.g. a URN) - the "official URI" would be decided by the API provider.

Other pages can then use the same `@id` to indicate that they contain (inofficial) information about the same API. Does that make sense and is it covered by schema.org semantics?

**Comment by thadguidry** [comment]
@LukasRos when folks need to compare 2 Things...yes, for disambiguation they can typically use the @id , the @url or even the http://schema.org/sameAs , or any combination of them.

**Comment by dberlind** [comment]
At ProgrammableWeb, for a great many reasons, we view the endpoint as the
most unique identifier.  It is honestly your best source of truth.

**Comment by westurner** [comment]
- https://schema.org/EntryPoint
  - https://schema.org/contentType
  - https://schema.org/encodingType
  - https://schema.org/httpMethod
  - https://schema.org/urlTemplate


- RPC
  - encodingType: ["json", "xml]
- SOAP
  - wsdlUrl
  - encodingType: xml

- "RESTful"
  - https://en.wikipedia.org/wiki/Representational_state_transfer#See_also
    - https://en.wikipedia.org/wiki/Overview_of_RESTful_API_Description_Languages
      - https://en.wikipedia.org/wiki/Overview_of_RESTful_API_Description_Languages#List_of_RESTful_API_DLs
      - 

- OData
- RSDL
- HATEOAS
  - https://en.wikipedia.org/wiki/HATEOAS
- Hydra
  - http://www.hydra-cg.com/spec/latest/core/



- WAMP
- WebSocket
- WebRTC
- Swagger / OpenAPI
  - type: APIFeature
    name: swagger
    version: 2.0
    url: https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md
    url: https://github.com/OAI/OpenAPI-Specification/blob/master/versions/1.2.md



 contentType:

- JSON: application/json, 
- JSONLD: application/ld+json
- XML
- CSV
- CSVW
- {mime/type}
- http://5stardata.info/en/

**Comment by hernan0216** [comment]
I worked in a prototype of a services catalog with this model called Minimal Service Model (MSM). I suggest that its recommendable take a look and uses schema concepts . http://kmi.github.io/iserve/latest/data-model.html

**Comment by MikeRalphson** [comment]
The chairs of the WebAPI Discovery Community Group are pleased to announce the first published draft of WADG0001 – a proposal to create an extension of the schema.org WebAPI type, to aid further discovery of machine- and human- readable API definitions and related documentation.

Feedback on this first draft, which is expected to undergo much change, is eagerly appreciated.

The report can be viewed at https://webapi-discovery.github.io/rfcs/rfc0001.html
Issues may be raised at https://github.com/webapi-discovery/rfcs/issues

**Comment by pvgenuchten** [comment]
Great initiative the webapi concept. Would it make sense to enable to link an api to one or more schema.org/dataset which may be exposed through an API (api's are commonly used to expose or author datasets).

In the iso19139 (geospatial data sharing) world we usually link iso19119 (services) to iso19115 (datasets).

**Comment by thadguidry** [comment]
@pvgenuchten This might be wise
property: providesDataset -> http://schema.org/Dataset

**Comment by westurner** [comment]
- C: http://schema.org/DataCatalog
  - P: http://schema.org/dataset r: http://schema.org/Dataset
    - http://schema.org/includedInDataCatalog

**Comment by thadguidry** [comment]
@westurner Yeah, just using our existing "http://schema.org/dataset" property is already generic already, and its definition would just need to be tweaked, and then can be reused on many other Types, including WebAPI.  +1  good find.

**Comment by LukasRos** [comment]
The statement that a WebAPI exposes a certain dataset is a very high-level connection between the API and the data, because it doesn't explain _how_ the API exposes the dataset. I'd say this should be included in the machine-readable API definition, for example as an extension to Swagger/OpenAPI.

**Comment by EricAxel** [comment]
My concern is around the new "documentation" property being proposed here (http://pending.schema.org/documentation). I suggest that the domain be expanded to include http://schema.org/Product - allowing for describing things such as owner manuals, assembly instructions, and manufacturer spec sheets.

**Comment by jpsherman** [comment]
I would also suggest something like 
documentUseCase 

- most documentation content has a purpose, such as installation, migration, assembly

documentChapterTitle

- most documentation has chapters that describe parts, assembly, troubleshooting

When it comes to software documentation I'd suggest a variation of "version" such as "releaseVersion" which would, when it comes to software, break down like this:

- product v1 = version

- product v1.2.3 = releaseVersion

overall, this is a fantastic idea, thank you for this.

**Comment by bardonelgen** [comment]
Maybe it could help to have a look at a practical example for a feasibility check:

Let’s say, a town’s authorities want to provide a digital interface to public services for their city-zens using Open311 and have a local hosting company provide the infrastructure on their behalf.

Now, you (as a loyal city dweller) come across a defective lamp post and would like to see it fixed.

You use the provided Open311 interface (via the API, using either a mobile app or a webpage) to notify the respective department of the incident (they probably receive an e-mail, mobile notification or simply a database entry).

Questions to be asked:

Who actually informs the local department ?   Is it you using the SERVICE ?  

Or does the service do this by itself (being the DIRECT per former of the action, which is actually out of your control…) ?  You, after all, only provided some raw data.

If, indeed, the SERVICE does: Wouldn’t it need to be an http://schema.org/agent ?  

But, as of now, these by definition need to be either human or an organisation ?

If the service could be an agent by itself, who will the be the http://schema.org/provider ?  

The local administration or the IT shop running the machines ?  Are THEY the ones sending the information then ?

Or still YOU yourself, as you – after all – initiated the process ?

And how about the generated output ?  Of course the API delivers a direct response to the user; but how about the output to 3rd parties (local authorities with this example) ?  

Is describing these outcomes out-of-scope for the API description or should it (or at least its typical outcome) be described with the API description for transparency purposes (a.k.a. “good practice”) ?

**Comment by acka47** [comment]
Thinking about how to add a link to an API and its documentation to the description of a website in the OER World Map (https://github.com/hbz/oerworldmap/issues/17#issuecomment-337519591), I noticed that we already use http://schema.org/ServiceChannel in a way where we can easily integrate API descriptions. Basically, we see a web site as a `Service` itself that can be accessed by different means (in the browser and via API). For example GitHub:

```json
{
   "name":"GitHub",
   "availableChannel":[
      {
         "@type":"ServiceChannel",
         "serviceUrl":"https://github.com",
         "availableLanguage":"en"
      },
      {
         "@type":"WebAPI",
         "serviceUrl":"https://api.github.com",
         "documentation":"https://developer.github.com/v3/"
      }
   ]
}
```

With this approach, `WebAPI` is used as a subclass of `ServiceChannel`. This probably isn't the way intended by schema.org but nonetheless I think it makes sense.

**Comment by rajbiswal** [comment]
Its a good idea to make APIs discoverable and adding the WebAPI to Schemas.org
We are looking for something like this for searching APIs based on context of an use-case or domain. At the moment there are multiple repositories of APIs but there is no standard to guide the structure for describing the APIs. Many don't even provide APIs to search the repository.

**Comment by monty241** [comment]
Would help us when support would be included for data models such as https://documentation.invantive.com/2017R2/exact-online-data-model/webhelp/index.html?invantive-eol-provider-exactonlinerest-codevalues-manufacturedbillofmaterialstatuses.htm

**Comment by MikeRalphson** [comment]
Is there a document which describes the process by which a type leaves pending?

**Comment by danbri** [comment]
@MikeRalphson https://github.com/schemaorg/schemaorg/blob/master/README.md has some notes on proposing schemas, and http://schema.org/docs/howwework.html provides a high level view of the whole thing, but we have not got a highly formalized approach for transferring things out of Pending into Core.

The most important considerations: "is anyone using (consuming) it, or at least likely to?" (if so please share details here), "does it add confusion or complication to schemas have already"  (e.g. in this case how does it related to [Entrypoint](http://schema.org/EntryPoint)), "are the term names sensible and not prone to misunderstanding, not overly general etc.".

**Comment by danbri** [comment]
I would like to see this move along and it seems useful. Here are a couple of things that came up when I was reviewing the topic (besides http://schema.org/EntryPoint from the /Action design):

1. How does it relate to the idea that a http://schema.org/Dataset has a http://schema.org/distribution that is a DataDownload? I suggest we amend /distribution to also note that WebAPI values are expected.

2. How does it relate to the idea that a SearchAction is a potentialAction on a WebSite? The idiom we use for this at Google is documented at https://developers.google.com/search/docs/data-types/sitelinks-searchbox - it captures something of the information needed to send a simple query to a site via HTTP/S GET URL encoding, e.g.

```<script type="application/ld+json">
{
  "@context": "http://schema.org",
  "@type": "WebSite",
  "url": "https://www.example.com/",
  "potentialAction": {
    "@type": "SearchAction",
    "target": "https://query.example.com/search?q={search_term_string}",
    "query-input": "required name=search_term_string"
  }
}
</script>```

**Comment by jaygray0919** [comment]
IOHO, `@EntryPoint` would be useful (and differentiating) if it is used to define REST and similar `endpoints`. One might argue that a `@Dataset` is content in an unspecified form. In contrast, an `@EntryPoint` might be content in a specified form (e.g. REST and JSON). Similarly, a SPARQL `endpoint` serves content, and expects the form (JSON, JSON-LD, RDF, TTL, N3, etc.) to be specified in the query ("potentialAction": ("@type": "SearchAction" ... )

One argument for adding `@EntryPoint` and differentiating from `@Dataset` ...

**Comment by dret** [comment]
maybe as one additional data point for this issue (which i think would be an excellent idea to pursue, schema.org currently is painfully lacking in terms of API descriptions): https://tools.ietf.org/html/draft-wilde-service-link-rel-06 is on the way to becoming an RFC and defines some link relations to identify resources that relate to web services (web APIs). this is clearly much more limited, but may be useful as input.

**Comment by AlasdairGray** [comment]
> 1. How does it relate to the idea that a http://schema.org/Dataset has a http://schema.org/distribution that is a DataDownload? I suggest we amend /distribution to also note that WebAPI values are expected.

In [Bioschemas](http://bioschemas.org) we would like to recommend a pattern for providing a link to an API for a Dataset. Is @danbri's suggestion here the expected approach?

**Comment by danbri** [comment]
Any more comments on https://webapi-discovery.github.io/rfcs/rfc0001.html ? Discussion so far has tended to focus on the datasets connection, which we should continue (and try to stay close to what DCAT are developing).

 But https://webapi-discovery.github.io/rfcs/rfc0001.html has a lot more to it. Is anyone ximplementing?

**Comment by dret** [comment]
On 2019-01-30 16:37, Dan Brickley wrote:
> Any more comments on 
> https://webapi-discovery.github.io/rfcs/rfc0001.html ? Discussion so far 
> has tended to focus on the datasets connection, which we should continue 
> (and try to stay close to what DCAT are developing).

just repeating my remarks about this replicating what 
https://tools.ietf.org/html/draft-wilde-service-link-rel-10 is doing 
(links to description and documentation).

adding to this, i'd say that adding "status" from that very same draft 
may be a good idea, too? many web APIs nowadays have status pages, and 
they are an interesting resource to know about.

should i raise these points as separate issues?

**Comment by bardonelgen** [comment]
Please excuse in case this is a duplicate, but does anyone already have 
an eye on the descriptions they use at Google use with their actionable 
e-mails ?

https://developers.google.com/gmail/markup/actions/declaring-actions

**Comment by danbri** [comment]
@bardonelgen  - my understanding is that the "webapi discovery" document linked above does try to use the same Schema.org Actions mechanism as the one used by Gmail.

@dret - thanks re https://tools.ietf.org/html/draft-wilde-service-link-rel-10 - we ought to express the mapping at least, if there is a clear equivalence.

And noted re "Status". Is there a subset we could start with, perhaps? e.g. webApiDefinitions, webApiActions, ...?

**Comment by westurner** [comment]
Should there be LDP (and Solid) Content-Types? Should content types contain the spec version, or no?

- https://en.wikipedia.org/wiki/Linked_Data_Platform
  - https://www.w3.org/TR/ldp/ (*1.0*)
- https://github.com/solid/solid-spec
  - https://github.com/solid/solid-spec/blob/master/api-rest.md
  - https://github.com/solid/solid-spec/blob/master/api-websockets.md

**Comment by dret** [comment]
On 2019-01-30 18:23, Dan Brickley wrote:
> @dret <https://github.com/dret> - thanks re 
> https://tools.ietf.org/html/draft-wilde-service-link-rel-10 - we ought 
> to express the mapping at least, if there is a clear equivalence.

it seems that the concepts of descriptions and documentation are pretty 
well aligned.

the difference is that link relations intentionally do not require any 
specific media types, but that's a different issue.

> And noted re "Status". Is there a subset we could start with, perhaps? 
> e.g. webApiDefinitions, webApiActions, ...?

subset of what? it seems that status might be a good fit for a starting 
point, in particular because it also is defined in the same spec.

**Comment by danbri** [comment]
@MikeRalphson @IvanGoncharov can you propose a pull request to update the version we have in Schema.org's Pending area, [based on your Community Group report](https://webapi-discovery.github.io/rfcs/rfc0001.html) (aka 'RFC', though I'm wary of calling it that as it may confuse [IETF folk](https://www.ietf.org/standards/rfcs/))

The files are 

* https://github.com/schemaorg/schemaorg/blob/master/data/ext/pending/issue-1423.rdfa (definitions)
and https://github.com/schemaorg/schemaorg/blob/master/data/ext/pending/issue-1423-examples.txt (examples)

I can help with mechanics. Essentially we need a label and description for each type and property, plus info that situates types in a hierarchy, and that relate types to properties.

**Comment by danbri** [comment]
Let me start trying to translate the doc's language into schema.org's machinery here -

* "versions (OPTIONAL array of Thing -> Property -> softwareVersion). It is RECOMMENDED that APIs be versioned using [semver]"
  * We already have https://schema.org.org/softwareVersion so we just declare softwareVersion has a domainIncludes of WebAPI 
  * We might also declare sofwareVersion a subPropertyOf the more general version property
* "entryPoints (OPTIONAL array of Thing -> Intangible -> EntryPoint)"
  * "extended with: responseContentTypes (OPTIONAL Text)"
   * Could we use encodingFormat instead?
* "license (OPTIONAL, CreativeWork or URL) - the license for the design/signature of the API"
  * this exists, http://schema.org/license
* "transport (enumerated Text: HTTP, HTTPS, SMTP, MQTT, WS, WSS etc)"
  * todo
* "apiProtocol (OPTIONAL, enumerated Text: SOAP, GraphQL, gRPC, Hydra, JSON API, XML-RPC, JSON-RPC etc)"
* "webApiDefinitions (OPTIONAL array of EntryPoints) containing links to machine-readable API definitions"
* "webApiActions (OPTIONAL array of potential Actions)"
* "The Content-Type(s) consumed by the WebAPI MAY be included in the EntryPoint.contentType property. The Content-Type(s) produced by the WebAPI MAY be included in the EntryPoint.responseContentType property.
    * All TODO

**Comment by fbastian** [comment]
I don't think `WebAPI` should be a `Service` subtype. `Service` in schema.org seems to refer to a system supplying a public need, not to the definition used in the computer science field.

See attributes such as `hoursAvailable`, `slogan`, or `serviceType`.

`WebAPI` should thus be a subtype of `SoftwareApplication`, in my opinion.

**Comment by westurner** [comment]
A WebAPI may be powered by various Software applications whilst still
having the same URL.

https://schema.org/termsOfService is a property of Thing > Intangible >
Service

https://schema.org/provider is a property of Service, and CreativeWork and
thus SoftwareApplication

https://schema.org/logo has a domain of {Service, Brand, Organization,
Place, Product}

**Comment by nickevansuk** [comment]
Hello @MikeRalphson @IvanGoncharov @danbri - this looks great, and is very applicable to our usecase.

We're standardising discovery and booking of physical activity through [OpenActive.io](https://www.openactive.io/) via a neighbouring [W3C Community Group for OpenActive](https://www.w3.org/community/openactive/), and are keen to use a profile of your spec as part of our more prescriptive [Dataset API Discovery spec](https://www.openactive.io/dataset-api-discovery/EditorsDraft/), and encourage adoption amongst our community.

We're providing a consistent mechanism for OpenActive [data publishers](https://status.openactive.io/) to make their datasets and APIs discoverable, and now have [many examples](https://opendata.fusion-lifestyle.com/OpenActive) of successfully published open datasets with schema.org markup.

We came across your work when we recently published [an API specification](https://www.openactive.io/open-booking-api/EditorsDraft/) that allows data users to interact with an openly published dataset through an authenticated API, and now need to incorporate discovery of that [`WebAPI`](https://schema.org/WebAPI) into our [discovery spec](https://www.openactive.io/dataset-api-discovery/EditorsDraft/).

Given all of the above:
1. @MikeRalphson @IvanGoncharov would you see a way for a [`Dataset`](https://schema.org/Dataset) to be linked to a [`WebAPI`](https://schema.org/WebAPI) within the current specification?
2. @danbri do you think Google would likely support a [`WebAPI`](https://schema.org/WebAPI) being embedded within a [`Dataset`](https://schema.org/Dataset) in JSON-LD markup - perhaps via a `schema:associatedWebAPI` property in the [`Dataset`](https://schema.org/Dataset)? If so we can propose it as part of this, if not I guess we'll need two different JSON-LD snippets on the [page](https://opendata.fusion-lifestyle.com/OpenActive)? Would that be supported?
3. @MikeRalphson @IvanGoncharov can we offer our resources to help with your current work at all? (e.g. we could create a PR for https://github.com/schemaorg/schemaorg/issues/1423#issuecomment-465274452?) Is there anything else we can do to help expedite the above into the schema.org pending extension?

**Comment by nickevansuk** [comment]
@MikeRalphson @IvanGoncharov @danbri so looking at the previous request to translate the doc's language into schema.org's machinery, while trying to stay close to what DCAT are developing, hopefully this is helpful:

Based on the [current WADG0001 WebAPI type extension specification](https://webapi-discovery.github.io/rfcs/rfc0001.html), considering [additional suggested amendments](https://github.com/schemaorg/schemaorg/issues/1423#issuecomment-465274452) applied, and the new [Data Catalog Vocabulary (DCAT) - Version 2](https://www.w3.org/TR/vocab-dcat-2/) please see the following proposed amends to make this schema.org-ready:

## Specification amendments

### `ActionableWebAPI` -> `pending:WebAPI`

The name `WebAPI` appears to have been already been agreed above?

### `entryPoints` -> `pending:endpointURL`

Can we use [DCAT v2 `endpointURL`](https://www.w3.org/TR/vocab-dcat-2/#Property:resource_conforms_to), the "The root location or primary endpoint of the service"?

### `contentType` -> `schema:encodingFormat`

`encodingFormat` in schema.org is currently used for MIME types, so is perhaps preferred to `contentType` for the _request_ (based on @danbri's [comment](https://github.com/schemaorg/schemaorg/issues/1423#issuecomment-465274452) above)? Otherwise we'd need to create a new property `accept` or `acceptFormat`, though this seems excessive...

However in 99% of cases the request and response content types would match? So do we need both request and response? Perhaps we just use `schema:contentType` for both?

Noting that `encodingType` in schema.org is defined as "The supported encoding type(s) for an EntryPoint request.",  so this isn't a good fit for the accept header of the request if "encoding" is related to HTTP `Accept-Encoding` (e.g. `gzip`)

### `responseContentTypes` -> `schema:contentType`

https://schema.org/contentType already defines this as the "supported content type(s) for an `EntryPoint` _response_", in line with the [HTTP definition of `Content-Type`](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html), so `schema:contentType` should cover the response.

### `license` -> `schema:license`

This exists, http://schema.org/license

### `transport` -> `pending:apiTransport`

`transport` is too generic for schema, suggest `apiTransport`

### `apiProtocol` -> `pending:conformsTo`

Can we use [DCAT v2 `conformsTo`](https://www.w3.org/TR/vocab-dcat-2/#Property:resource_conforms_to) here, which makes the property more generally applicable? And recommend URLs here as canonical descriptions of the different protocols? E.g. `https://jsonapi.org/format/1.0/`, `https://grpc.io/`, `http://www.hydra-cg.com/spec/latest/core/`

### `webApiDefinitions` -> `pending:endpointDescription`

Can we bring this in line with [DCAT v2 `endpointDescription`](https://www.w3.org/TR/vocab-dcat-2/#Property:data_service_endpoint_description) and use the term `endpointDescription`? Using `EntryPoint` and `contentType` to allow the different machine-readable descriptions to be differentiated?

### `webApiActions` -> `schema:potentialAction`

Is just standard `schema:potentialAction`? The `potentialAction` being defined on a `WebAPI` class makes it clear they are "webApiActions"?

Suggest no need to define the [`name` values](https://webapi-discovery.github.io/rfcs/rfc0001.html#webapiactions) as `Action` `name` is not designed to be used to reference a vocabulary. At least not at this stage? A consuming application can just render the names to the user as links, hence "apiAuthentication" becomes "API authentication requirements", etc?

### `pending:accessService`

The [DCAT v2 term `dcat:accessService`](https://www.w3.org/TR/vocab-dcat-2/#Property:distribution_access_service) links the `dcat:Dataset` to the `dcat:DataService`. Given that `pending:WebAPI` subclasses `schema:Service`, could we just bring this in as `pending:accessService` with domain `schema:Dataset` and range `pending:WebAPI`? This meets the needs of the folks above who have been discussing the `Dataset` <-> `WebAPI` link.

Note we could also add [DCAT v2 term `dcat:servesDataset`](https://www.w3.org/TR/vocab-dcat-2/#Property:data_service_serves_dataset) as the reverse of this, domain `pending:WebAPI` and range `schema:Dataset`.

### `versions` -> `schema:version`

The following version-related properties exist in schema.org. It appears that `versions` in the current spec is describing the API `version` (e.g. "Member Web Services 1.3") rather than the `softwareVersion` (e.g. "LeisureSoft 2.0"), especially in the case of on-premise software that exposes APIs, so perhaps `schema:version` seems more appropriate?

- http://schema.org/schemaVersion
Indicates (by URL or string) a particular version of a schema used in some CreativeWork. For example, a document could declare a schemaVersion using an URL such as http://schema.org/version/2.0/ if precise indication of schema version was required by some application.
Used on `CreativeWork`

- https://schema.org/version
The version of the CreativeWork embodied by a specified resource.
Used on `CreativeWork`

- http://schema.org/assemblyVersion
Associated product/technology version. e.g., .NET Framework 4.5.
Used on `APIReference`

- http://schema.org/softwareVersion
Version of the software instance.
Used on `SoftwareApplication`


## Example

```
{
  "@context": "http://schema.org/",
  "@type": "Dataset",
  ...
  "accessService": {
    "@type": "WebAPI",
    "name": "Google Knowledge Graph Search API",
    "description": "The Knowledge Graph Search API lets you find entities in the Google Knowledge Graph. The API uses standard schema.org types and is compliant with the JSON-LD specification.",
    "documentation": "https://developers.google.com/knowledge-graph/",
    "termsOfService": "https://developers.google.com/knowledge-graph/terms",
    "logo": "https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png",
    "license": "https://creativecommons.org/licenses/by/3.0/",
    "provider": {
      "@type": "Organization",
      "name": "Google Inc.",
      "contactPoint": [
        {
          "@type": "ContactPoint",
          "name": "Google",
          "url": "https://google.com"
        }
      ],
    },
    "version": [
      "1.0.0"
    ],
    "endpointURL": [
      {
        "@type": "EntryPoint",
        "url": "https://example.com/api/openbooking/",
        "contentType": "application/json"
      }
    ],
    "apiTransport": "HTTPS",
    "conformsTo": [
      "https://jsonapi.org/format/1.0/"
    ],
    "endpointDescription": [
      {
        "@type": "EntryPoint",
        "contentType": "application/vnd.oai.openapi+json;version=2.0",
        "url": "https://www.openactive.io/open-booking-api/1.0/swagger.json"
      },
    ],
    "potentialAction": [
      {
        "@type": "ConsumeAction",
        "name": "API Client Registration",
        "target": "https://exampleforms.com/get-me-an-api-access-key"
      }
    ]
  }
}
```

## Feedback

What do you think? We're happy to do this in a PR if that's helpful to move forward? Perhaps then we leave these in Pending until DCAT v2 makes it out of CR, but at least get some implementation experience in the mean time?

**Comment by jeffrogers** [comment]
Are there provisions to describe whether the API is publicly consumable or whether it requires authentication and authorization?

**Comment by shibumi** [comment]
Hi, everybody. I am a junior researcher and I am highly interested in this topic.
I just have a few questions: Did you thought about adding/using concepts of the OWL-S or WSMO for your WebAPI Ontology for schema.org?

I know, both concepts are from 2004. But there were people who thought about the same concepts.
For example they've have introduced the concept of "Inputs, Outputs, Preconditions, Effects" for a service. Input and Output refer to a a webAPI (you know, the normal JSON request and response...) + semantic description of both. Preconditions and Effects, however, refer to an action in the physical word.. like "your creditcard needs to be charged for using this service" and effects like:"Your creditcard is charged now"..

what do you think?

**Comment by westurner** [comment]
What would be necessary to support either or both standards with schema.org
RDFS?

**Comment by shibumi** [comment]
I strongly suggest not to support it. However it would make sense to adopt concepts from both standards. The standards are de facto dead (nobody is using it, except a small research group bubble maybe). No tools exist for these standard and XML/WSDL seems pretty outdated, but the some of the concepts are great.

**Comment by shibumi** [comment]
So what needs to be done to finish this project? I have registered in the W3 community and I would like to help.

I've found also a paper that doesn't look so different: http://www.semantic-web-journal.net/system/files/swj1144.pdf

Do you have any paper about this webapi type submission?

**Comment by smrgeoinfo** [comment]
Some other efforts along these lines: 
EarthCube Resource Registry, [API resource type](https://github.com/earthcubearchitecture-ecresourcereg/infomodel/blob/ba3823bf015363fc7793c0724772b2317848cd51/SyntaxForTabularResourceDescriptions.md#service-interchange-format-interfaceapi)

[Discussion of Machine actionable Links](http://usgin.github.io/usginspecs/ContentModelForLinks.htm), somewhat dated, but covers a lot of the content information that needs to be accounted.

Example schema.org JSON-LD:
[CHORDS](https://github.com/earthcubearchitecture-ecresourcereg/infomodel/blob/c67500b5da73c2faf330072b4e7f897f03b160ad/examples/Interface-ChordsWebAPIv2-SDO.json)
[IRIS data](https://github.com/earthcubearchitecture-ecresourcereg/infomodel/blob/4b7cdaf408522d03edc0f096708103cc17a451e7/examples/Service-IRIS-fsdnEvent-JSON.json)
[Template for service description](https://github.com/earthcubearchitecture-ecresourcereg/infomodel/blob/4b7cdaf408522d03edc0f096708103cc17a451e7/examples/ECRRTemplateSDO-Service-v2.JSON), using [potential action](https://github.com/ESIPFed/science-on-schema.org/blob/a1f2d30cf660204e8d93b58a695e455e2f13ea53/guides/Dataset.md#accessing-data-through-a-service-endpoint)

There's a lot of interest in this in the science data community!

**Comment by smrgeoinfo** [comment]
see also https://github.com/schemaorg/schemaorg/issues/2340
https://github.com/schemaorg/schemaorg/issues/2342
https://github.com/earthcubearchitecture-project418/p419dcatservices
https://www.w3.org/TR/vocab-dcat-2/#Class:Data_Service

**Comment by smrgeoinfo** [comment]
From the perspective of EarthScience data, we're dealing with LOTS of data with varying degrees of granularity.  Packaged 'datasets' that can be downloaded as files (simple sdo:downloadURL) are common, but volumetrically minor compared to the volume of data becoming available via API based services (for accessing legacy data) and dataFeeds for accessing real-time data. What we need are conventions for how to generate a high level record describing e.g.

- a monitoring site that has multiple sensors, with datafeeds from each sensor that are time series for which the sampling interval, and temporal coverage are important. Granular access to the actual time series required for research is via a webAPI that supports requests for site, sensor type, time window, aggregation...etc.

- a geochemical database that includes data from ~100K  sites, multiple samples from a site, and various analytical results from each sample. Metadata needs to describe the webAPI that enables requests for particular sampled features, analytes, sampling time, analysis method...

-- a geothermal well database that has ~million wells, with various logs and observation data for each well.  Metadata needs to describe the webAPI that enables request for particular observations, from particular depths, particular formations, sampling time, observation method.

you get the picture...   Some kind of API description that indicates what parameters can be specified for filtering, what kind of request input or responst output format can be specified -- not just JSON, XML, RDF, but what is the schema, what vocabularies are used, what is the profile for the interchange format.  [URI templates ](https://tools.ietf.org/html/rfc6570)are an obvious approach, but there alow needs to be[ content negotiation](https://www.w3.org/TR/dx-prof-conneg/) possible on the requests , and there is currently no mechanism via URI templates to indicate content negotiation options....

**Comment by Sohamd00** [comment]
We at [Sprinkle Data](https://www.sprinkledata.com/), use multiple APIs to extract the data.

**Comment by smrgeoinfo** [comment]
@Sohamd00 if those APIs were to be exposed via schema.org descriptions for indexing by some aggregator, and use by other client applications to automate access to Sprinkle data, how would you build those descriptions with schema.org vocabulary?

**Comment by Sohamd00** [comment]
@smrgeoinfo Sprinkle does extract and import data using API keys. But for building those descriptions with schema.org we need to have a detailed understanding.

**Comment by nickevansuk** [comment]
Just updating the thread here with some progress that has been made on this:
- https://webapi-discovery.github.io/rfcs/rfc0001.html has been updated based on the finalised DCAT v2 and the suggestions above, and its language translated into schema.org's machinery
- We've created a PR based on this: https://github.com/schemaorg/schemaorg/pull/2635

**Comment by nickevansuk** [comment]
So #2635 is now set for [inclusion in the next schema.org release](https://github.com/schemaorg/schemaorg/issues/2648), however there is one outstanding issue that we're seeking consensus on...

It would be great to get your opinion on https://github.com/webapi-discovery/rfcs/issues/10, in the context of [this draft](https://webapi-discovery.github.io/rfcs/rfc0001.html).

Any other feedback welcome of course!

**Comment by westurner** [comment]
Examples of how to represent GraphQL, SPARQL, LDP, and SOLID APIs would tremendously improve likelihood of adoption of this very useful method for API discovery.

When I link people to https://schema.org/WebAPI and reference #example2, I shouldn't expect them to do much more than copy and paste and change values.

**Comment by nickevansuk** [comment]
Just noting this currently blocked on @danbri response on #2635. We have consensus on this next step otherwise.

**Comment by ThisIsMissEm** [comment]
Should this also cover a services' privacy policy and acceptable use policy documents? (not just the terms of service)

**Comment by AlasdairGray** [comment]
Is the `documentation` property only for linking to the API definition or can it be used to link to other types of documentation and for other types of resources.

Could it be used to link a dataset to a web page that provides more detailed documentation about the dataset than you find on the landing page, e.g. for UniProt the landing page would point to uniprot.org but the documentation could link to the help page https://www.uniprot.org/help.

**Comment by phtyson** [comment]
> Is the `documentation` property only for linking to the API definition or can it be used to link to other types of documentation and for other types of resources.
> 
> Could it be used to link a dataset to a web page that provides more detailed documentation about the dataset than you find on the landing page, e.g. for UniProt the landing page would point to uniprot.org but the documentation could link to the help page https://www.uniprot.org/help.

Agree. "documentation" is very generic and the domain should be any Product or Software. If a specific api documentation property is wanted, the name and domain specification should say so.

**Comment by westurner** [comment]
How do the WebAPI use cases overlap with OpenAPI, JSON-RPC, and MCP use cases?

Use Cases / Examples for :WebAPI :

- :Service 
  https://schema.org/Service
  - :provider (d:Service, d:Action,)
    https://schema.org/provider
- :Action
  https://schema.org/Action
  > An action performed by a direct agent and indirect participants upon a direct object. Optionally happens at a location with the help of an inanimate instrument. The execution of the action may produce a result. Specific action sub-type documentation specifies the exact expectation of each argument/role.
  - :agent (d:Action,) and :provider (d:Service,) are similar to W3C PROV from what I recall
  - :agent
    https://schema.org/agent
- .
- Dataset:
  http://schema.org/Dataset
- DataCatalog:
  http://schema.org/DataCatalog
- .
- EntryPoint (d: Action.target)
  https://schema.org/EntryPoint

## WebAPI
- [Thing](https://schema.org/Thing) > [Intangible](https://schema.org/Intangible) > [Service](https://schema.org/Service) > [WebAPI](https://schema.org/WebAPI)
- Issues: is:issue is:open WebAPI: https://github.com/schemaorg/schemaorg/issues?q=is%3Aissue+is%3Aopen+WebAPI

## OpenAPI
- Wikipedia: https://en.wikipedia.org/wiki/OpenAPI_Specification
- Spec: https://spec.openapis.org/oas/latest.html
- .
- Spec > Security Considerations:  https://spec.openapis.org/oas/latest.html#security-considerations
- .
- List of implementations of OpenAPI: https://tools.openapis.org/
- GitHub topic: OpenAPI:   https://github.com/topics/openapi (5,088)
- /? "openapi" "rdf" https://www.google.com/search?q=%22openapi%22+%22rdf
- /? inurl:awesome site:github.com openapi : https://www.google.com/search?q=inurl%3Aawesome+site%3Agithub.com+openapi


# JSON-RPC
- Wikipedia:  https://en.wikipedia.org/wiki/JSON-RPC
- Spec : https://www.jsonrpc.org/specification (v2.0 (2010))
- Web: https://www.jsonrpc.org/
- .  
- GitHub topic: json-rpc: https://github.com/topics/json-rpc (657)
- Github topic: jsonrpc:  https://github.com/topics/jsonrpc (226)
 - /? inurl:awesome site:github.com json-rpc : https://www.google.com/search?q=inurl%3Aawesome+site%3Agithub.com+json-rpc
  - https://github.com/shanejonas/awesome-json-rpc#interface-description

Also,

## MCP: Model Context Protocol

- Spec:  https://spec.modelcontextprotocol.io/specification/
- Src: https://github.com/modelcontextprotocol/specification
- Web : https://modelcontextprotocol.io/
- Docs:  https://github.com/modelcontextprotocol/docs/tree/main/docs/
- Docs: https://modelcontextprotocol.io/docs/
- .
- MCP is a JSON-RPC 2.0 spec
- Docs:  https://github.com/modelcontextprotocol/docs/blob/main/docs/concepts/architecture.mdx
- Docs: https://modelcontextprotocol.io/docs/concepts/architecture
- .
- List of Model Context Protocol Servers: https://github.com/modelcontextprotocol/servers
- GitHub topic: https://github.com/topics/mcp (913)
- mcp-servers-hub List of MCP servers: https://github.com/apappascs/mcp-servers-hub
- /? inurl:awesome site:github.com json-rpc : https://www.google.com/search?q=inurl%3Aawesome+site%3Agithub.com+json-rpc
  - awesome-mcp-ervers list of MCP servers: https://github.com/punkpeye/awesome-mcp-servers
- .
- /? "MCP" "rdf":  https://www.google.com/search?q=%22MCP%22+%22rdf
- /? "MCP" "rdf"JSON-RPC: https://www.google.com/search?q=%22MCP%22+%22RDF%22+%22JSON-RPC%22

## W3C PROV
- Wikipedia: https://en.wikipedia.org/wiki/PROV_(Provenance)
- Activity, Agent, Entity
- Spec: https://www.w3.org/TR/prov-overview/
- PROV Primer (W3C TR): 2. Intuitive overview of PROV: https://www.w3.org/TR/prov-primer/#intuitive-overview-of-prov
- .
- /? inurl:awesome site:github.com provenance https://www.google.com/search?q=inurl%3Aawesome+site%3Agithub.com+provenance
- awesome-fair > Provenance tracking: https://github.com/Materials-Data-Science-and-Informatics/awesome-fair#provenance-tracking
- awesome-semantic-web#misc > git2prov: https://github.com/semantalytics/awesome-semantic-web#misc
  - https://github.com/IDLabResearch/Git2PROV

## Scientific workflows with provenance
- The W3C PROV specifications model provenance from processes that could be scientific workflows.
- e.g. DVC:
  - src: https://github.com/iterative/dvc
  - docs: https://github.com/iterative/dvc#quick-start
  - .
  - DVC is an example of a tool for reproducible experiments with provenance metadata and version control on schema:Datasets that could be retrieved from a schema:WebAPI.
  - .
  - GitHub Topic dvc: https://github.com/topics/dvc (458)
    - IDE Support for the dvc workflow:
      - https://github.com/iterative/vscode-dvc
  - https://dvc.org/doc/user-guide/project-structure
    - dvc.yaml
      - YAML-LD?
    - .dvc
    - .dvcignore
  - /? "dvc" "w3c prov"
    https://www.google.com/search?q="dvc"+"w3c+prov"
    - When a prov:Agent A prov:used an prov:Entity to generate another prov:Entity that prov:wasGeneratedBy an prov:Activity that prov:wasAssociatedWith said prov:Agent A,
      and a prov:Entity was a schema:Dataset retrieved from schema:WebAPI ,
      how should PROV and WebAPI and the API method/query result be linked with RDF?
      - This is beyond the immediate scope of WebAPI,
        but what about schema:Action.target schema:Entrypoint?
      - What is the relation between schema:EntryPoint and schema:WebAPI?
        - https://schema.org/EntryPoint

**Comment by lisad** [comment]
If this is to include privacy policies it needs to be multi-valued - a service may well have a terms of service, a privacy policy, and its organization's privacy policy.

**Comment by westurner** [comment]
- https://schema.org/Service
  - https://schema.org/termsOfService
  - [ ] :privacyPolicy

- https://schema.org/Organization
  - [ ] :privacyPolicy

- Newer API types:
  - [ ] MCP
  - [ ] MCP-B

Are there additional types of WebAPIs?

**Comment by westurner** [comment]
Additional 'types' of WebAPIs: 
- MCP: https://modelcontextprotocol.io/specification/latest
- WebMCP: https://docs.mcp-b.ai/introduction
  - src: https://github.com/WebMCP-org
- UTCP: https://www.utcp.io/about/RFC
  - https://www.utcp.io/protocols
  - src: https://github.com/universal-tool-calling-protocol
- ACP:
  - https://www.ibm.com/think/topics/agent-communication-protocol
- A2A: https://a2a-protocol.org/latest/specification/
  - src: https://github.com/a2aproject/A2A
  - https://a2a-protocol.org/latest/roadmap/
    - `.well-known/agent-card.json`
      - https://a2a-protocol.org/latest/specification/#5-agent-discovery-the-agent-card
        > A2A Servers MUST make an Agent Card available. The Agent Card is a JSON document that describes the server's identity, capabilities, skills, service endpoint URL, and how clients should authenticate and interact with it. Clients use this information for discovering suitable agents and for configuring their interactions.
      - https://a2a-protocol.org/latest/topics/agent-discovery/
      - https://a2a-protocol.org/latest/tutorials/python/3-agent-skills-and-card/#agent-card

**Comment by westurner** [comment]
- d: schema.org/EntryPoint
  > [Thing](https://schema.org/Thing) > [Intangible](https://schema.org/Intangible) > [EntryPoint](https://schema.org/EntryPoint)
  > An entry point, within some Web-based protocol.
  > - [actionApplication](https://schema.org/actionApplication) 	r:[SoftwareApplication](https://schema.org/SoftwareApplication) 	An application that can complete the request. Supersedes - [application](https://schema.org/application).
  > - [actionPlatform](https://schema.org/actionPlatform) 	r:[DigitalPlatformEnumeration](https://schema.org/DigitalPlatformEnumeration)  or [Text](https://schema.org/Text)  or [URL](https://schema.org/URL)
  >   The high level platform(s) where the Action can be performed for the given URL. To specify a specific application or operating system instance, use actionApplication.
  > - [contentType](https://schema.org/contentType) 	r:[Text](https://schema.org/Text) 	The supported content type(s) for an EntryPoint response.
  > - [encodingType](https://schema.org/encodingType) 	r:[Text](https://schema.org/Text) 	The supported encoding type(s) for an EntryPoint request.
  > - [httpMethod](https://schema.org/httpMethod) 	r:[Text](https://schema.org/Text) 	An HTTP method that specifies the appropriate HTTP method for a request to an HTTP EntryPoint. Values are capitalized strings as used in HTTP.
  > - [urlTemplate](https://schema.org/urlTemplate)  r:[Text](https://schema.org/Text) 	An url template (RFC6570) that will be used to construct the target of the execution of the action.

**Comment by westurner** [comment]
EntryPoint.urlTemplate ... OpenSearch URL templates ... Swagger Path Templates

OpenSearch:
- wikipedia: https://en.wikipedia.org/wiki/OpenSearch_(specification)
- spec: https://github.com/dewitt/opensearch/blob/master/opensearch-1-1-draft-6.md
- https://github.com/dewitt/opensearch/blob/master/opensearch-1-1-draft-6.md#the-url-element :
  > ```html
  > <Url type="text/html"
  >      template="http://example.com/search?q={searchTerms}&amp;pw={startPage?}" />
  > ```

RFC 6570: URI Template:
- spec: https://www.rfc-editor.org/rfc/rfc6570.html :
```
http://example.com/~{username}/
http://example.com/dictionary/{term:1}/{term}
http://example.com/search{?q,lang}
```

OpenAPI / Swagger:
- spec: https://swagger.io/specification/
- https://swagger.io/specification/#path-templating 
- https://swagger.io/specification/#paths-object
- ```
  /pets/{petId}
  /pets/mine
  ```
- ```json
  "url": "https://{username}.gigantic-server.com:{port}/{basePath}",
  ```

**Comment by lindangulopez** [comment]
Hi all,

Following [ongoing discussions in the CodeMeta community](https://github.com/codemeta/codemeta/discussions/335), I’d like to propose extending the `documentation` property so that it can reference any relevant `CreativeWork`, not only API definitions or URLs.

Many software and dataset projects rely on diverse forms of documentation—README files, tutorials, manuals, help pages, build guides, technical notes, etc. Restricting `documentation` to `APIReference`/`URL` makes it difficult to represent these resources without creating multiple new, highly specific properties. Allowing `documentation` to accept `CreativeWork | URL` would support far broader use cases while remaining consistent with schema.org modeling patterns.

This need has been raised by several community members, including @AlasdairGray and others who have asked whether `documentation` could link to dataset help pages, software guides, and other non-API documentation, etc.

In CodeMeta, we are considering introducing a temporary `documentation` property (`CreativeWork | URL`) for CodeMeta v4 to accommodate this gap, but we would rather rely directly on schema.org’s property than duplicating it. Extending the expected type upstream would resolve multiple long-standing issues around documentation and instructions.

Would schema.org be open to expanding the expected type of `documentation` to include `CreativeWork`?

Thank you for considering this request. 

Kind regards,
Linda

**Comment by MatthiasWiesmann** [comment]
Could you make a concrete proposal on what needs to be added, e.g. types and attributes?
Would a Documentation super-class / cop-type make sense?

**Comment by lindangulopez** [comment]
Hi @MatthiasWiesmann ,

Thanks for following up with @dgarijo.  We discussed this in the CodeMeta community today, and we agree that it makes sense to keep things specific to avoid pushback.

Therefore, we suggest using the existing TechArticle class for technical documentation, such as manuals, guides, and software-related documentation. This approach eliminates the need to introduce a new superclass, such as TechnicalWorks.

Specifically, this would mean:

- Keeping "documentation" as a property on relevant items (e.g., SoftwareSourceCode, Dataset).
- Allowing "documentation" to reference "TechArticle | URL."
- Reusing an existing, well-understood schema.org type rather than adding a new one.

This approach:

- Is more specific than "CreativeWork."
- Avoids introducing new classes
- Covers most real-world documentation use cases, such as API documentation, manuals, technical guides, and help pages.
It is already being adopted informally by the CodeMeta community. See the discussion converging toward "TechArticle" at https://github.com/codemeta/codemeta/discussions/453#discussioncomment-15277664.

If this sounds reasonable, we would be happy to align CodeMeta v4 with it and provide concrete examples to support the change.

Thanks again for the thoughtful [discussion](https://github.com/codemeta/codemeta/discussions/335)

Kind regards, 
Linda

**Comment by MatthiasWiesmann** [comment]
Thanks, would it also make sense to have a reverse property for TechArticle, something like documents:

**Comment by lindangulopez** [comment]
yes indeed @MatthiasWiesmann : there are many uses cases.

**Comment by fthobe** [comment]
@gmackenz solved by https://schema.org/WebAPI and documented if you approve #4674 with the changes you requested.

**Comment by gmackenz** [comment]
> [@gmackenz](https://github.com/gmackenz) solved by https://schema.org/WebAPI and documented if you approve [#4674](https://github.com/schemaorg/schemaorg/pull/4674) with the changes you requested.

Minor issue in the usage of """ for only some comment structures in the PR--not to strong on my RDFa formatting but does that denote something special for the comments that have more than one bracketing " or is just harmless.

**Comment by fthobe** [comment]
@gmackenz perfectly valid for multi line as it seems:

https://www.w3.org/TR/turtle/#literals

" and ' can be used interchangably, but you can use only the one not encapsulating the text string inside the text string itself without escaping. Triple are used for paragraphs.

Example single ticks: `'He said "Hello" to his friend!'`
Example Multiline: 

```
'''
He said
"Hello"
to his friend.
'''
```

**Comment by fthobe** [comment]
@MatthiasWiesmann there's a property documentation inside with a swagger as example. I remember you are doing some work on software. Do you
- Have any preference on linking the documentation? 
- at this point shouldn't we also provide a privacy link analogue to terms of service as a property on Organization and classes describing software or web services of any kind?

Reflecting on it I would actually expand the scope of documentation to also cover product and software application.

**Comment by fthobe** [comment]
@gmackenz do you think we can merge this and close two issues?

**Comment by gmackenz** [comment]
> [@gmackenz](https://github.com/gmackenz) do you think we can merge this and close two issues?

Sure, I think the example and extensions look fine to me. I think the code change for the python script really should be a separate issue/review.

**Comment by fthobe** [comment]
@gmackenz yes, I will single it out

**Comment by rebeccaringuette** [comment]
Any updates on when the change will be implemented?

**Comment by fthobe** [comment]
Hey Rebecca, I will take a look this weekend and wrap it up.

**Comment by dgarijo** [comment]
It looks like https://schema.org/WebAPI was added, so this issue may be closed

**Comment by westurner** [comment]
- webapi-discovery W3C CG: https://github.com/webapi-discovery
- RFC: WADG0001 WebAPI type extension https://webapi-discovery.github.io/rfcs/rfc0001.html :
  > This extension has been created based on real-world use of metadata in the [OpenAPI], [RAML], [API-Blueprint], [JSON-home] and [apis.json] specifications

**Comment by fthobe** [comment]
@gmackenz we still need the examples I have open in a PR to pickup again
